controller/resource_roles_blackbox_test.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/fabric8-services/fabric8-auth/authorization/role"
+	. "github.com/fabric8-services/fabric8-auth/controller"
+	"github.com/fabric8-services/fabric8-auth/gormtestsupport"
+
+	testsupport "github.com/fabric8-services/fabric8-auth/test"
+	"github.com/goadesign/goa"
+	"github.com/satori/go.uuid"
+
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type TestResourceRolesRest struct {
+	gormtestsupport.DBTestSuite
+}
+
+func (s *TestResourceRolesRest) SetupSuite() {
+	s.DBTestSuite.SetupSuite()
+}
+
+func (rest *TestResourceRolesRest) SecuredControllerWithIdentity(identity account.Identity) (*goa.Service, *ResourceRolesController) {
+	svc := testsupport.ServiceAsUser("Resource-roles-Service", testsupport.TestIdentity)
+	roleAssignmentModelService := models.NewRoleAssignmentModelService(rest.DB, rest.Application)
+	roleAssignmentService := authorization.NewRoleAssignmentService(roleAssignmentModelService, rest.Application)
+	return svc, NewResourceRolesController(svc, rest.Application, roleAssignmentService)
+}
+
+func TestRunResourceRolesRest(t *testing.T) {
+	suite.Run(t, &TestResourceRolesRest{DBTestSuite: gormtestsupport.NewDBTestSuite()})
+}
+
+func (rest *TestResourceRolesRest) TestListAssignedRolesOK() {
+
+	// Create a resource of the inbuilt resource type
+	// Create a role for that resource type
+	// Create two assignments for that role.
+
+	resourceOwner := testsupport.TestIdentity2
+	err := testsupport.CreateTestIdentityAndUserInDB(rest.DB, &resourceOwner)
+	require.NoError(rest.T(), err)
+
+	areaResourceType, err := rest.Application.ResourceTypeRepository().Lookup(rest.Ctx, "openshift.io/resource/area")
+	require.NoError(rest.T(), err)
+
+	resourceRef, err := testsupport.CreateTestResource(rest.Ctx, rest.DB, *areaResourceType, "SpaceR")
+	require.NoError(rest.T(), err)
+
+	roleRef, err := testsupport.CreateTestRole(rest.Ctx, rest.DB, *areaResourceType, "collab")
+	require.NoError(rest.T(), err)
+
+	var createdIdentityRoles []role.IdentityRole
+
+	identityRoleRef, err := testsupport.CreateTestIdentityRole(rest.Ctx, rest.DB, *resourceRef, *roleRef)
+	require.NoError(rest.T(), err)
+	require.NotNil(rest.T(), identityRoleRef)
+	createdIdentityRoles = append(createdIdentityRoles, *identityRoleRef)
+
+	identityRoleRef2, err := testsupport.CreateTestIdentityRole(rest.Ctx, rest.DB, *resourceRef, *roleRef)
+	require.NoError(rest.T(), err)
+	require.NotNil(rest.T(), identityRoleRef2)
+	createdIdentityRoles = append(createdIdentityRoles, *identityRoleRef2)
+
+	svc, ctrl := rest.SecuredControllerWithIdentity(testsupport.TestIdentity)
+	_, returnedIdentityRoles := test.ListAssignedResourceRolesOK(rest.T(), rest.Ctx, svc, ctrl, resourceRef.ResourceID)
+	require.Len(rest.T(), returnedIdentityRoles.Data, 2)
+	require.True(rest.T(), rest.checkExists(*identityRoleRef, returnedIdentityRoles, false))
+	require.True(rest.T(), rest.checkExists(*identityRoleRef2, returnedIdentityRoles, false))
+}
```

Comment:
In any layer (controller, service, model service) create more than expected assigned roles in tests. Check that only expected roles are returned. Just to make sure we do not return not relevant roles. I would probably put most covering tests to the service because the service may be re-used in different controllers. But I'm not sure... Don't want to duplicate all tests in all layers... But it's OK to have some overlap in the tests in different layers.

Reply:
Added tests in controller, and a lot more detailed tests in the service layer https://github.com/fabric8-services/fabric8-auth/pull/322/commits/d70fc33c99735c506f1654d9c199e45615f05a28